Offer with oracle event

da-kami opened the pull request:
fixes #193

da-kami committed:
Introduce `OracleEventId` wrapper type

da-kami committed:
Maker publishes `OracleEventId` as part of offer

The latest announcement known to the maker's cfd actor is used for extracting the event id.

da-kami committed:
Taker uses latest announcements from oracle to start contract setup

Opted not to implement an additional message to fetch a specific announcement from the Oracle actor because it would have made the oracle actor unnecessarily complex (We would have to split it into different functionality by trait).
The taker cfd actor holds a sliding window of all the announcements received from the oracle actor.

da-kami committed:
Maker uses latest announcements from oracle in contract setup

thomaseizinger commented on daemon/src/oracle.rs:
```diff
@@ -200,9 +201,9 @@ fn event_url(datetime: OffsetDateTime) -> String {
     /// Identifier for an oracle event.
     ///
-    /// Doubles up as the path of the URL for this event i.e.
+    /// Bubbles up as the path of the URL for this event i.e.
```
I think this is meant to say _doubles_.

thomaseizinger replied:
The expression "To double up as something" means it serves two purposes :)

thomaseizinger commented on daemon/src/maker_cfd.rs:
```diff
@@ -99,7 +103,26 @@ impl Actor {
+            latest_announcement.id
+        } else {
+            bail!("Cannot create order because no announcement from oracle")
+        };
```
`.context` on an `Option` does exactly this.

thomaseizinger commented on daemon/src/routes_maker.rs:
```diff
@@ -95,15 +95,14 @@ pub async fn post_sell_order(
+        })
         .await
-        .expect("actor to always be available");
+        .map_err(|e| status::BadRequest(Some(e.to_string())))?;
```
I know you just took it over but I really think this should be a 500. This failing has nothing to do with the user's request.

thomaseizinger commented on daemon/src/taker_cfd.rs:
```diff
@@ -60,8 +60,8 @@ pub struct Actor {
     setup_state: SetupState,
-    latest_announcement: Option<oracle::Announcement>,
-    _oracle_actor: Address<oracle::Actor<Actor, monitor::Actor<Actor>>>,
+    latest_announcements: Option<oracle::Announcements>,
```
Can we make this a `HashMap` instead? Sounds like the better data structure if we are going to look-up by event ID.

thomaseizinger replied:
Just make sure to actively clear it on every update, otherwise we have a memory leak here.

da-kami replied:
Went for keeping the array in the oracle and map + replace on every update in the other actors - ok?

thomaseizinger commented on daemon/src/maker_cfd.rs:
```diff
@@ -64,8 +64,8 @@ pub struct Actor {
     setup_state: SetupState,
-    latest_announcement: Option<oracle::Announcement>,
-    _oracle_actor: Address<oracle::Actor<Actor, monitor::Actor<Actor>>>,
+    latest_announcements: Option<oracle::Announcements>,
```
You can use a `BTreeMap` here, that is a sorted `HashMap` so you can grab the last one but still look up by ID.

thomaseizinger reviewed (approved):
Thanks!

Some suggestions for using better data structures but nothing blocking.

luckysori commented on daemon/src/model.rs:
```diff
@@ -105,3 +105,6 @@ pub struct WalletInfo {
 }
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct OracleEventId(pub String);
```
❓ Would it make sense to put this type in `oracle` and name it `Id`?

da-kami replied:
This type is now used in various modules - it is also serialized/deserialized as part of the `Order`. We could be cleaner here, and map it to a specific type for serialization in `to_sse_event`. I admit I was too lazy for that. But I would keep it in `model` and treat it similar to e.g. `Usd` which is also ussed throughout the codebase but directly serialized for the UI.

thomaseizinger replied:
Would be nice to parse this one into its semantic parts at some point :)

luckysori commented on daemon/src/model/cfd.rs:
```diff
@@ -90,11 +90,22 @@ pub struct Order {
+
+    /// The id of the event to be used for price attestation
+    ///
+    /// The maker includes this into the Order based on the Oracle's next event.
```
🔧 Technically it's not the _next_ event, but the event 24 hours from now IIRC.

luckysori commented on daemon/src/taker_cfd.rs:
```diff
@@ -184,18 +184,23 @@ impl Actor {
+            .0
+            .iter()
+            .find(|announcement| announcement.id == cfd.order.oracle_event_id)
+            .context("Order's announcement not found in list of current oracle announcements")?;
```
👍 So instead of asking the oracle actor for a specific announcement we assume that the announcement in the order we get from the maker is within our `latest_announcements` state? It should work and it's definitely simpler to implement it this way. 

Perhaps we can go back an introduce a dedicated message later on.

luckysori commented on daemon/src/taker_cfd.rs:
```diff
@@ -363,8 +368,7 @@ impl Actor {
         &mut self,
         announcements: oracle::Announcements,
     ) -> Result<()> {
-        tracing::debug!("Updating latest oracle announcements");
```
😭 Why don't you like my arbitrary, useless debug statements?

da-kami replied:
hahaha, I think I actually removed that one by accident - in the maker I kept it 🤦‍♂️

luckysori commented on daemon/src/maker_cfd.rs:
```diff
@@ -338,18 +344,23 @@ impl Actor {
+            .0
+            .iter()
+            .find(|announcement| announcement.id == cfd.order.oracle_event_id)
+            .context("Order's announcement not found in list of current oracle announcements")?;
```
❓ I can't see the types here, so maybe I'm missing something, but can't you just use the `cfd.order.oracle_event_id` directly when sending the `oracle::MonitorEvent` message to the oracle actor? The maker shouldn't need to verify anything about an offer _they_ created.

luckysori replied:
Okay, ignore this. You need to get the full announcement to get the `nonce_pks`.

da-kami replied:
We need the actual `Announcement` to retrieve the `nonce_pks` after we send the message to the monitor actor. 
In the `Order` we only have the `OracleEventId`. I think it is safer to only trigger the monitoring once we know we also have the announcement (containing the `nonce_pks`), that's why I first retrieve the announcement and only if that succeeded go on with triggering monitoring.

luckysori commented on daemon/src/maker_cfd.rs:
```diff
@@ -486,7 +497,7 @@ impl Actor {
         announcements: oracle::Announcements,
     ) -> Result<()> {
         tracing::debug!("Updating latest oracle announcements");
```
🔧 To be consistent you could also delete this useless debug statement.

da-kami replied:
hahaha, and there it is :D

luckysori reviewed (approved):
Looks good to me, just some little fixes.

da-kami committed:
`BTreeMap` for announcements in cfd actors

Replaces the option on ever update by mapping the oracle's announcement array into a BTreeMap.

da-kami committed:
Remove debug message

thomaseizinger commented on daemon/src/maker_cfd.rs:
```diff
@@ -543,7 +540,14 @@ impl Actor {
+                .0
+                .iter()
+                .map(|announcement| (announcement.id.clone(), announcement.clone()))
+                .collect(),
```
Nice use of `collect`!

thomaseizinger commented on daemon/src/taker_cfd.rs:
```diff
@@ -63,7 +63,7 @@ pub struct Actor {
     monitor_actor: Address<monitor::Actor<Actor>>,
     setup_state: SetupState,
-    latest_announcements: Option<oracle::Announcements>,
+    latest_announcements: Option<BTreeMap<OracleEventId, oracle::Announcement>>,
```
Not sure if we need to semantically differentiate between `Option::None` and an empty map, so I think we could get rid of the `Option`.

thomaseizinger reviewed (approved):
Thanks!
